### PR TITLE
Add VQA V2.0 and Visual Dialog V0.9.

### PR DESCRIPTION
**Pull request by jiasenlu**
This pull request contains 
1: add vqa_v2 task loader.
2: add VisDial_v0.9 task loader.
3: disable download COCO image as default 
4: move COCO image path to "--download_path" 

To test the new added task

For VQA 2.0: python examples/display_data.py -t vqa_coco2014_v2 --download-path 'path_to_COCO_img'

For Visual Dialog: python examples/display_data.py -t visdial

Currently, the Visual Dialog inherit from the default "DialogTeacher" class. There is no placeholder for the additional image information. The "DialogData" class with the format [(x, y, r, c), new_episode?], could we extend this to  [(x, y, r, c, i), new_episode?] where "i" is some optional additional information such as image_id? If so, I can send another pull request to modify that.

### CODE REVIEW

**Comment by alexholdenmiller** [comment]
Thanks Jiasen!! I think we'd like to keep the dataset downloading the images automatically (to datapath)--I know it's large, but users should only need to do it once anyways. Everything in downloads is also automatically downloaded (currently just the Memnn github repo).

As far as DialogTeacher/DialogData, I think we want to write a new one instead that includes some of the data loading from your vqa_coco2014 (v1) and prepares us to handle the preprocessing options that we talked about before. We don't want image_id as part of the action/observation dictionary--just images themselves--so we want to include the image loading as part of that data class.

File: parlai/tasks/vqa_coco2014/agents.py [hunk 1]
```diff
@@ -36,7 +35,7 @@ def _path(opt):
     annotation_path = os.path.join(opt['datapath'], 'VQA-COCO2014',
         annotation_suffix + '_annotations.json')
 
-    image_path = os.path.join(opt['datapath'], 'VQA-COCO2014', img_suffix)
```
Comment (jiasenlu) [inline parlai/tasks/vqa_coco2014/agents.py]: I see. Then I think we might not want the image stay in 'VQA-COCO2014' folder, other task such as Visual Dialog may also use the COCO image. How about put the image under `COCO-IMG` ? Then, multiple task can share the image data.

File: parlai/tasks/vqa_coco2014_v2/agents.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        qa = self.ques['questions'][self.episode_idx]
+        question = qa['question']
+        image_id = qa['image_id']
+        # question_id = qa['question_id']
+
+        if self.datatype != 'test':
+            anno = self.annotation['annotations'][self.episode_idx]
+            answers = [ans['answer'] for ans in anno['answers']]
+        else:
+            answers = ['fake_answer']
```
Comment (jiasenlu) [inline parlai/tasks/vqa_coco2014_v2/agents.py]: ok, thanks. will change this.

File: parlai/tasks/vqa_coco2014/agents.py (hunk 1 above)
Comment (alexholdenmiller) [inline parlai/tasks/vqa_coco2014/agents.py]: Sounds great! Yeah giving it the most general name for the data file is perfect, and then multiple tasks can depend on it (and it won't rebuild it if it's already there).

File: parlai/core/dialog_teacher.py [hunk 3]
```diff
@@ -265,7 +265,7 @@ def get(self, episode_idx, entry_idx=0):
                 table['reward'] = entry[2]
                 if len(entry) > 3:
                     table['label_candidates'] = entry[3]
-                    if len(entry) > 4 and not opt.get('no_images', False):
+                    if len(entry) > 4 and not self.opt.get('no_images', False):
```
Comment (alexholdenmiller) [inline parlai/core/dialog_teacher.py]: ah good catch thank you

File: parlai/tasks/vqa_coco2014_v2/agents.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        if self.datatype != 'test':
+            print('loading: ' + annotation_path)
+            with open(annotation_path) as data_file:
+                self.annotation = json.load(data_file)
+
+        self.len = len(self.ques['questions'])
+
+class DefaultTeacher(OeTeacher):
+    pass
```
Comment (alexholdenmiller) [inline parlai/tasks/vqa_coco2014_v2/agents.py]: does v2 have a multiple-choice version?

**Comment by jaseweston** [comment]
yes, we need to have at least a simple example of visualizing the images..

On Thu, May 11, 2017 at 11:37 AM, Alexander Miller <notifications@github.com

File: parlai/tasks/vqa_coco2014_v2/agents.py (hunk 4 above)
Comment (jiasenlu) [inline parlai/tasks/vqa_coco2014_v2/agents.py]: no, VQA v2.o doesn't have the multiple-choice now.

**Comment by alexholdenmiller** [comment]
simple example which prints ascii version of images (after `pip install asciimatics`), or we could write our own similar one

```python
import sys
from asciimatics.renderers import ImageFile
img = ImageFile(sys.argv[1], height=30)
print(img)
```

**Comment by jiasenlu** [comment]
Yeah, I think we should do that. Maybe we can add this in a separate pull request?

**Comment by alexholdenmiller** [comment]
Just pushed this to get that fix in

**Pull request merged**